### ISSUE TITLE
Quickfix does not get populated in non-continuous mode

### ISSUE THREAD

**Issue by michamos** [lervag/vimtex#275]
Hi,

I just switched from Latex-Box to vimtex, and all together it seems like a very nice plugin.
The only issue I have so far is that the log does not get loaded when using non-continuous mode (although compilation itself works correctly).
As far as I understand, the reason for that is that `cfile`/`cgetfile` get called in the callback that latexmk only triggers on continuous compilation.

Maybe for non-continuous mode, it would be simpler to leverage the vim compiler support by defining `makeprg`?

These are the relevant settings in my vimrc:
```
let g:vimtex_latexmk_continuous=0
let g:vimtex_latexmk_background=1
let g:vimtex_latexmk_options="-pdflatex='pdflatex -synctex=1 \\%O \\%S'"
```

**Comment by lahwaacz** [comment]
If I remember correctly, `makeprg` is run synchronously...

**Comment by lervag** [comment]
@michamos As @lahwaacz mentions, the problem is that `makeprg` runs synchronously, which blocks Vim while compiling. That is simply not acceptable.

In LaTeX-Box, the non-continuous command is complex in order to support the callback. I prefer to keep things simple (or at least try hard to), and so I chose to not support this feature.

If you have any good ideas for how to support this feature, I would be interested in knowing it. It has to work on Linux, OSX and Windows, and it should be simple (relatively, at least).

**Comment by lervag** [comment]
Note, @michamos, that you may always check your document yourself after compilation with `\le`.

**Comment by michamos** [comment]
Indeed, `makeprg` runs synchronously, which is a general problem with vim. Fortunately, there exist plugins such as [vim-dispatch](https://github.com/tpope/vim-dispatch) which provides several strategies for running `makeprg` asynchronously according to the user's environment.

So one could do the following:
1. define `makeprg` in the `latexmk` compiler plugin
2. use `:Make` for non-continuous foreground compiling, fall back to `:make` if `vim-dispatch` not available
3. use `:Make!` for non-continuous background compiling, fall back to current behavior (optional, depending on the amount of simplification if it is removed, which is probably small as continuous-mode needs to be supported anyway). Errors can then be loaded with `:Copen` (manually or automatically by `vimtex`).

This goes well with the vimtex philosophy of being simple and delegating functionality to other plugins instead of reinventing the wheel (e.g. Ultisnips for `imap`s).

> Note, @michamos, that you may always check your document yourself after compilation with \le.

Thanks, I had not noticed this.

**Comment by lervag** [comment]
@michamos Yes, I agree, you could define `makeprg` and couple this with `vim-dispatch`. However, I think this should be possible to do on a user level, and thus it does not need to be part of vimtex.

That is, you should be able to utilize the `vimtex#latexmk#errors()` in a custom `makeprg` setting. Although, I actually think a more sensible thing to do is to create a simple makefile and to have `makeprg=make`. For instance, this should work as a simple makefile:

```make
default:
	-latexmk -verbose -pdf
	gvim --remote-expr "vimtex#latexmk#errors()"
```

Note here the leading `-`. This is important to ensure that the remote expression is called even when latexmk fails.

The bottom line is that the "simple" one shot compilation is relatively easy to implement by the user. However, it is relatively (perhaps even very) hard to implement in a general way that will work on all platforms and for most users.
